### PR TITLE
Add player view model

### PR DESCRIPTION
**Pull request by jstarry95**
Refactored AddPlayerViewController so that it conforms to MVVM. Do this end, I added a class AddPlayerViewModel with the functionality required for the AddPlayerViewController. Please let me know if my approach achieves the required level of abstraction, fam.

### CODE REVIEW

File: BoardBank/MonoBank/AddPlayerViewController.swift [hunk 1]
```diff
@@ -95,7 +85,7 @@ class AddPlayerViewController: UITableViewController, UITextFieldDelegate, UICol
   }
   
   func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
-    selectedToken = indexPath.item
+    addPlayerViewModel.selectedToken = indexPath.item
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewController.swift]: Make selectedToken private. Have the view controller call a function to update the selected token instead

File: BoardBank/MonoBank/AddPlayerViewController.swift [hunk 2]
```diff
@@ -84,8 +74,8 @@ class AddPlayerViewController: UITableViewController, UITextFieldDelegate, UICol
   
   func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
     let tokenCell = collectionView.dequeueReusableCell(withReuseIdentifier: "tokenCell", for: indexPath) as! TokenCollectionViewCell
-    let isSelected = indexPath.item == selectedToken
-    let tokenName = isSelected ? tokens[indexPath.item] + "_filled" : tokens[indexPath.item]
+    let isSelected = indexPath.item == addPlayerViewModel.selectedToken
+    let tokenName = isSelected ? addPlayerViewModel.tokens[indexPath.item] + "_filled" : addPlayerViewModel.tokens[indexPath.item]
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewController.swift]: The token name should be gotten from the view model. Create a function that takes in isSelected, and returns a string.

File: BoardBank/MonoBank/AddPlayerViewController.swift [hunk 3]
```diff
@@ -15,15 +15,12 @@ class AddPlayerViewController: UITableViewController, UITextFieldDelegate, UICol
...
-  
-  let tokens = ["cannon", "car", "cat", "dog", "hat", "horse", "iron", "money", "ship", "shoe", "thimble", "wheelbarrow"]
-  var selectedToken = 0
+  let addPlayerViewModel = AddPlayerViewModel()
   
   override func viewDidLoad() {
-    currencySymbolLabel.text = bankManager.currencySymbol
-    balanceTextField.text = String(bankManager.defaultBalance)
-    balanceTextField.placeholder = String(bankManager.defaultBalance)
+    currencySymbolLabel.text = addPlayerViewModel.getBankManagerDefaults().currencySymbol
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewController.swift]: There should be a function to get the currency symbol.

File: BoardBank/MonoBank/AddPlayerViewController.swift [hunk 4]
```diff
@@ -15,15 +15,12 @@ class AddPlayerViewController: UITableViewController, UITextFieldDelegate, UICol
...
+    balanceTextField.text = String(addPlayerViewModel.getBankManagerDefaults().defaultBalance)
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewController.swift]: There should be a function to get the default balance.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+//  AddPlayerViewModel.swift
+//  BoardBank
+//
+//  Created by Consultant on 7/23/19.
+//  Copyright © 2019 Richard Neitzke. All rights reserved.
+//
+
+import UIKit
+
+class AddPlayerViewModel {
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Final class

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  let bankManager = BankManager()
+  
+  let tokens = ["cannon", "car", "cat", "dog", "hat", "horse", "iron", "money", "ship", "shoe", "thimble", "wheelbarrow"]
+  var selectedToken = 0
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: All these properties should be private.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  
+  func getBankManagerDefaults() -> (currencySymbol: String, defaultBalance: Int32) {
+    let currencySymbol = self.bankManager.currencySymbol
+    let defaultBalance = self.bankManager.defaultBalance
+    let resultTuple = (currencySymbol: currencySymbol, defaultBalance: defaultBalance)
+    
+    return resultTuple
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Why a tuple over a struct?

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  }
+  
+  func addPlayer(nameText: String, balanceText: String) {
+    let strippedBalance = balanceText.components(separatedBy: CharacterSet.decimalDigits.inverted).joined()
+    let name = nameText.isEmpty ? "Player" : nameText
+    let balance = Int32(strippedBalance) == nil ? self.getBankManagerDefaults().defaultBalance : Int32(strippedBalance)!
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Don't force unwrap. For this case, you could use nil coalescing.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+      return 44
+    }
+  }
+  
+  func getHeightForHeader(for section: Int) -> CGFloat {
+    if section == 0 {
+      return 25
+    } else {
+      return 5
+    }
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Use ternary over this approach.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  }
+  
+  
+}
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Blank line at end of file.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  
+  func getHeightForHeader(for section: Int) -> CGFloat {
+    if section == 0 {
+      return 25
+    } else {
+      return 5
+    }
+  }
+  
+
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Too many blank lines here.

File: BoardBank/MonoBank/BankManager.swift [hunk 12]
```diff
@@ -29,19 +29,19 @@ final class BankManager {
   }
   
   /// Default balance when adding a player
-  var defaultBalance: Int {
+  var defaultBalance: Int32 {
     get {
-      return userDefaults.value(forKey: "defaultBalance") as? Int ?? 1500
+      return userDefaults.value(forKey: "defaultBalance") as? Int32 ?? 1500
```
Comment (prrenfrew) [inline BoardBank/MonoBank/BankManager.swift]: Why was this changed to Int32?

File: BoardBank/MonoBank/BankManager.swift [hunk 13]
```diff
@@ -29,19 +29,19 @@ final class BankManager {
...
       userDefaults.set(newValue, forKey: "defaultBalance")
     }
   }
   
   /// Amount which the user can quickly add in the player menu
-  var quickAddAmount: Int {
+  var quickAddAmount: Int32 {
     get {
-      return userDefaults.value(forKey: "quickAddAmount") as? Int ?? 200
+      return userDefaults.value(forKey: "quickAddAmount") as? Int32 ?? 200
```
Comment (prrenfrew) [inline BoardBank/MonoBank/BankManager.swift]: Why was this changed to Int32?

File: BoardBank/MonoBank/BankManager.swift [hunk 14]
```diff
@@ -74,8 +74,9 @@ final class BankManager {
     numberFormatter.currencySymbol = self.currencySymbol
   }
   
+
   @discardableResult
-  func addPlayer(name: String, balance: Int, token: Token) -> Player? {
+  func addPlayer(name: String, balance: Int32, token: Token) -> Player? {
```
Comment (prrenfrew) [inline BoardBank/MonoBank/BankManager.swift]: Why was this changed to Int32?

File: BoardBank/MonoBank/BankManager.swift (hunk 12 above)
Comment (jstarry95) [reply BoardBank/MonoBank/BankManager.swift]: The balance in CoreData is Int32. I figured uniformity would be preferred.

File: BoardBank/MonoBank/BankManager.swift (hunk 13 above)
Comment (jstarry95) [reply BoardBank/MonoBank/BankManager.swift]: See Above

File: BoardBank/MonoBank/BankManager.swift (hunk 14 above)
Comment (jstarry95) [reply BoardBank/MonoBank/BankManager.swift]: See Above

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+final class AddPlayerViewModel {
+  private let bankManager = BankManager()
+  
+  private let tokens = ["cannon", "car", "cat", "dog", "hat", "horse", "iron", "money", "ship", "shoe", "thimble", "wheelbarrow"]
+  
+  private var selectedToken = 0
+  
+  func getCurrencySymbol() -> String {
+    return self.bankManager.currencySymbol
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: Don't use self where not required.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    let tokenRawValue = tokens[selectedToken]
+    guard let token = Token(rawValue: tokenRawValue) else { return }
+    
+    bankManager.addPlayer(name: name, balance: balance, token: token)
+  }
+  
+  func getRowHeight(for section: Int, and row: Int, tokenCellSize: CGSize) -> CGFloat {
+    if section == 1 && row == 0 {
+      // iPhone has two rows, iPad has one row
+      let rowsHeight = UIDevice.current.userInterfaceIdiom == .phone ? tokenCellSize.height*2: tokenCellSize.height
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: There should be a space on either side of the multiply asterisk `*`.

File: BoardBank/MonoBank/AddPlayerViewModel.swift [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+      return 44
+    }
+  }
+  
+  func getHeightForHeader(for section: Int) -> CGFloat {
+    return section == 0 ? 25 : 5
+  }
+  
+  func getToken(at index: Int) -> String {
+    return tokens[index]
```
Comment (prrenfrew) [inline BoardBank/MonoBank/AddPlayerViewModel.swift]: What if the index is an invalid index? Right now it just crashes.

**Comment by prrenfrew** [comment]
This view model needs unit tests. Otherwise looks pretty good.
